Afbase/sync1.1 part 1

afbase opened the pull request:
This is going to be the first of many different PRs related to Sync1.1  

I think the PR strategy would be to do many small, reviewable ones instead of one big one.

As I make PRs in the future, I will address and document what I am fixing and how it conforms to the typescript implementation in #66.

## Files Changed

1. `rsky-lexicon/src/com/atproto/sync.rs`
2. `rsky-pds/src/actor_store/mod.rs`
3. `rsky-pds/src/actor_store/repo/mod.rs`
4. `rsky-pds/src/actor_store/repo/types.rs` (new file)
5. `rsky-pds/src/apis/com/atproto/admin/delete_account.rs`
6. `rsky-pds/src/apis/com/atproto/admin/update_account_handle.rs`
7. `rsky-pds/src/apis/com/atproto/server/delete_account.rs`
8. `rsky-pds/src/apis/com/atproto/sync/subscribe_repos.rs`
9. `rsky-pds/src/lexicon/lexicons.toml`
10. `rsky-pds/src/sequencer/events.rs`
11. `rsky-pds/src/sequencer/mod.rs`

## Summary

1. **Replaced Tombstone and Handle events with a new Sync event type**:
   - Removed `TombstoneEvt` and `HandleEvt` types
   - Added a new `SyncEvt` and `TypedSyncEvt` structure
   - `SubscribeReposTombstone` replaced with `SubscribeReposSync`

2. **Updated synchronization data structure**:
   - Added a new `SyncEvtData` structure containing repository metadata (CID, rev, blocks)
   - Added functionality to get sync event data via `get_sync_event_data()` method

3. **Modified account handling**:
   - Updated account deletion logic to remove tombstone event sequencing
   - Updated handle update logic, removing explicit handle update sequencing
   - Added a TODO comment for an incomplete branch in account handle updates

4. **Updated the event sequencing system**:
   - Modified the event deserialization logic to handle the new event types
   - Updated the request sequence range function to use the new event types

5. **Modified repository subscription mechanism**:
   - Updated the `subscribe_repos` function to handle the new sync events
   - Updated lexicon schemas in TOML configuration

The changes appear to be streamlining the repository synchronization process by consolidating multiple event types into a more comprehensive sync event that includes repository block data directly, rather than using separate events like tombstones and handle updates.

## Related Issues

Relates to #66 and the PRs:

- https://github.com/bluesky-social/atproto/commit/6e382f67aa73532efadfea80ff96a27b526cb178
- https://github.com/bluesky-social/atproto/pull/3585

## Changes

- [x] Feature implementation
- [x] Other (please specify): conforms to sync1.1

## Checklist
- [ ] ~I have tested the changes (including writing unit tests).~
- [x] I confirm that my implementation aligns with the [canonical Typescript implementation](https://github.com/bluesky-social/atproto) and/or [atproto spec](https://atproto.com/specs/atp)
- [x] ~I have updated relevant documentation.~ As I make PRs in the future, I will address and document what I am fixing and how it conforms to the typescript implementation in #66.
- [x] I have formatted my code correctly
- [ ] ~I have provided examples for how this code works or will be used~

afbase commented:
I would call out that this should have close eyes on the changes during the review of this.  I could totally see feedback for other things to incorporate in the PR, etc.

afbase commented:
@rudyfraser 

Do you know if there is anything in rsky that corresponds to `packages/ozone/src/lexicon` in the atproto typescript repo?  I'm going through https://github.com/bluesky-social/atproto/pull/3391 with my PR to see what changes I need to add.

afbase commented:
Will continue with conforming to https://github.com/bluesky-social/atproto/pull/3612/files#diff-e4cbcaefd701028a584c82c21d6d6ab680f39e9da11e260d3aebb348054c5e92

rudyfraser commented:
> @rudyfraser 
> 
> 
> 
> Do you know if there is anything in rsky that corresponds to `packages/ozone/src/lexicon` in the atproto typescript repo?  I'm going through https://github.com/bluesky-social/atproto/pull/3391 with my PR to see what changes I need to add.

Not a direct thing but worth checking rsky-lexicon. That'd be the only place.

afbase commented:
> [!NOTE]  
> https://github.com/blacksky-algorithms/rsky/pull/74 should be merged before this one

rudyfraser commented:
@afbase is this one ready for review? I merged #74 
Also including [Bryan Newbold's post](https://bsky.app/profile/bnewbold.net/post/3lktb6hki2c2g):

> for any devs working on sync v1.1 induction firehose:
> 
> I added verification functionality to goat, which should be pretty close to new spec and new relay. you can run it against a > PDS, eg:
> 
> 'goat firehose --relay-host wss://pds.robocracy.org --verify-mst --verify-basic'
> 
> and it will log problems

afbase commented:
> @afbase is this one ready for review? I merged #74 Also including [Bryan Newbold's post](https://bsky.app/profile/bnewbold.net/post/3lktb6hki2c2g):
> 
> > for any devs working on sync v1.1 induction firehose:
> > I added verification functionality to goat, which should be pretty close to new spec and new relay. you can run it against a > PDS, eg:
> > 'goat firehose --relay-host wss://pds.robocracy.org --verify-mst --verify-basic'
> > and it will log problems

@rudyfraser  I'm still working on this one. also I need help testing this one when it's ready as i don't have rsky-pds stood up for testing anywhere.

afbase commented:
@rudyfraser 

A few things I'm trying to reconcile for [sync 1.1](https://github.com/bluesky-social/proposals/tree/main/0006-sync-iteration)

## Firehose #commit validation

>include additional metadata and MST blocks in firehose #commit messages to enable per-commit MST validation ("operation inversion" or "inductive firehose")

AFAICT, atproto repo doesn't implement this just quite yet but i think the pseudo-code might look something like:

```rust
pub fn validate_commit_via_inversion(
    commit_data: &CommitDataWithOps, 
    prev_data: &Cid
) -> Result<bool> {
    // 1. Start with the current state in blocks
    let mut mst = MST::load_from_blocks(&commit_data.blocks)?;
    
    // 2. Invert all operations
    for op in &commit_data.ops {
        match op.action {
            CommitEvtOpAction::Create => {
                // Invert create -> delete
                mst.delete(&op.path)?;
            },
            CommitEvtOpAction::Update => {
                // Invert update -> update to previous value
                if let Some(prev) = &op.prev {
                    mst.update(&op.path, prev)?;
                } else {
                    return Err(anyhow::anyhow!("Missing prev CID for update operation"));
                }
            },
            CommitEvtOpAction::Delete => {
                // Invert delete -> create with previous value
                if let Some(prev) = &op.prev {
                    mst.create(&op.path, prev)?;
                } else {
                    return Err(anyhow::anyhow!("Missing prev CID for delete operation"));
                }
            }
        }
    }
    
    // 3. After inverting all operations, the tree root should match prevData
    let final_root = mst.get_root_cid()?;
    Ok(final_root == *prev_data)
}
```

## Staying Synchronized: #sync event, auto-repair, and account status

Still need to implement this.  also AFAICT this isn't quite implemented in atproto.  


## Lexicons and listReposByCollection

I have some notes for implementing listReposByCollection endpoint.  I'll continue working on that.

Also, this is more of a general question on lexicons but my thinking is we might want to have some automation in place to manage lexicons, yes?

afbase commented:
The following comment is with regard to commit `1f119cb`:

> [!IMPORTANT]  
> This comment has two sections: "Implementation of `com.atproto.sync.listReposByCollection`" and "Future Improvements".  Future Improvements is lengthy and the opentelemetry details is long but worth a follow to get an idea of future direction and telemetry practice.


> [!WARNING]  
> @rudyfraser  I would like a review of these changes before I commit any more on this branch


# Implementation of `com.atproto.sync.listReposByCollection`

1f119cb implements the `com.atproto.sync.listReposByCollection` endpoint as specified in the [ATProto API specification](https://docs.bsky.app/docs/api/com-atproto-sync-list-repos-by-collection). This endpoint allows clients to retrieve all DIDs that have records of a specified collection, with efficient cursor-based pagination.

1. This implementation is over the PDS whereas the bluesky implementation is done over a relay (e.g. https://relay1.us-east.bsky.network/xrpc/com.atproto.sync.listReposByCollection?collection=app.bsky.feed.generator).  I presume implementing this over a repo is okay but please let me know if this is what we should be doing.

## Changes Overview

### 1. `rsky-pds/src/apis/com/atproto/sync/list_repos_by_collection.rs` (New File)

The main implementation file with several key components:

- **Response Type**: Created `ListReposByCollectionOutput` struct matching the API specification
- **Collection Validation**: Implemented `NsidCollection` wrapper with NSID validation.  @rudyfraser @TheRipperoni if either of you review this, would you mind validating this part as well?  I'm worried I'm going to have this evaluating things "backwards".
- **Cursor Mechanism**: Implemented efficient cursor-based pagination using base64-encoded cursors
- **Database Query Optimization**: Leveraged the collection index for efficient querying
- **Error Handling**: Implemented comprehensive error handling that conforms to the API standard
- **Reusable Components**: Created reusable types that implement Rocket's request guards

#### Database Query Optimization

The implementation takes advantage of the `record_did_collection_idx` index (on the `collection` column) in the database:

1. Main query filters by `collection` first, which utilizes the index
3. For cursor-based pagination, we use a compound condition:
   ```sql
   WHERE collection = $1 
   AND (did > $2 OR (did = $2 AND CTID::text::point[0]::int8 > $3))
   ```
4. This allows efficient continuation from a previous page without scanning the entire table
5. We use raw SQL for the cursor case to have more control over the query plan
6. Ordering by `did` ensures consistent pagination results

### 2. `rsky-pds/src/apis/com/atproto/sync/mod.rs`

Updated to include the new module:

```rust
pub mod list_repos_by_collection;
```

### 3. `rsky-pds/src/lib.rs`

Registered the new endpoint in the Rocket application:

```rust
com::atproto::sync::list_repos_by_collection::list_repos_by_collection,
```

### 4. `rsky-pds/src/apis/mod.rs`

Added `std::fmt::Display` implementation for `ApiError`, which improves error handling and logging throughout the application:

```rust
impl std::fmt::Display for ApiError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ApiError::InvalidRequest(msg) => write!(f, "InvalidRequest: {}", msg),
            // ... other variants
        }
    }
}
```

## Implementation Details

### Cursor Design

- The cursor encodes both the last seen DID and a row identifier to ensure accurate continuation
- Base64 encoding is used for the cursor to make it URL-safe
- Proper error handling for malformed cursors

### Type Safety Improvements

- Created `NsidCollection` type to validate collection NSIDs
- Implemented `FromParam` trait for `NsidCollection` to validate router parameters
- Created `CursorWrapper` to handle cursor data with proper Rust type safety

### Query Efficiency

- Uses `DISTINCT` to avoid duplicate DIDs over the records table
- Orders results by `did` for deterministic pagination
- Uses the existing collection index for performance
- Properly specifies SQL types in Diesel bindings
- Optimized for both cursor and non-cursor cases

### Error Handling

- Comprehensive validation for input parameters
- Proper API error responses that conform to the specification
- Detailed logging of internal errors while keeping public error messages appropriate

# Future Improvements

## Compound Index Implementation

The current implementation uses the existing `record_diollection_idx` index on the `collection` column. While this works well for filtering by collection, a compound index on `(collection, did)` would significantly improve query performance, especially for pagination scenarios.

### Implementation Plan for Compound Index

1. **Migration File Changes**

```sql
-- Add to an upcoming migration file
CREATE INDEX record_collection_did_idx
    ON pds.record(collection, did);
```

3. **Diesel Query Optimization**

Once the compound index is in place, we could update the Diesel query to take full advantage of it:

```rust
// Current implementation with standard index
let results: Vec<String> = RecordSchema::record
    .select(RecordSchema::did)
    .filter(RecordSchema::collection.eq(collection.as_str()))
    .distinct()
    .order_by(RecordSchema::did.asc())
    .limit(limit + 1)
    .load(conn)?;
    
// Enhanced implementation with compound index hint
let results: Vec<String> = RecordSchema::record
    .select(RecordSchema::did)
    .filter(RecordSchema::collection.eq(collection.as_str()))
    // For PostgreSQL, we can use more explicit index hints if needed
    // This can be done via raw SQL fragments in Diesel
    .order_by(RecordSchema::did.asc())
    .limit(limit + 1)
    .load(conn)?;
```

4. **Cursor Optimization**

With a compound index, we could simplify the cursor logic and make it more efficient:

```rust
// Using the compound index directly in cursor queries
if let Some(cursor_data) = &cursor_data {
    let raw_query = format!(
        "SELECT DISTINCT did FROM pds.record 
        -- This WHERE clause is now covered by a single compound index
        WHERE collection = $1 AND did > $2
        ORDER BY collection, did ASC
        LIMIT $3"
    );
    
    sql_query(raw_query)
        .bind::<Text, _>(collection.as_str())
        .bind::<Text, _>(&cursor_data.did)
        .bind::<diesel::sql_types::BigInt, _>(limit + 1)
        .load::<DidResult>(conn)?
}
```

5. **Performance Metrics**

We could implement performance metrics collection to measure the impact:

```rust
use std::time::{Instant, Duration};

async fn inner_list_repos_by_collection(
    collection: NsidCollection,
    limit: Option<i64>,
    cursor_data: Option<CursorData>,
    db: DbConn,
) -> Result<Json<ListReposByCollectionOutput>, ApiError> {
    let start = Instant::now();
    
    // Query execution code...
    
    let query_duration = start.elapsed();
    // Log or collect metrics
    tracing::debug!(
        "Collection query completed in {:?} ms, rows: {}", 
        query_duration.as_millis(),
        repos.len()
    );
    
    // Rest of function...
}
```

## Other Future Enhancements

### 1. Advanced Cursor Encoding

We could enhance the cursor encoding to support more complex pagination scenarios:

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EnhancedCursor {
    last_seen_id: i64,
    did: String,
    // Additional fields for more complex cursor functionality
    timestamp: chrono::DateTime<chrono::Utc>,
    collection_hash: u64, // Hash of collection for verification
}
```

### 2. Comprehensive Testing

Add thorough unit and integration tests:

```rust
#[cfg(test)]
mod tests {
    use super::*;
    
    #[rocket::async_test]
    async fn test_nsid_collection_validation() {
        // Test validation logic
    }
    
    #[rocket::async_test]
    async fn test_cursor_encoding_decoding() {
        // Test cursor functionality
    }
    
    #[rocket::async_test]
    async fn test_pagination_with_mock_db() {
        // Test pagination with mock database
    }
    
    // Integration tests with a test database
    #[rocket::async_test]
    async fn test_with_real_database() {
        // Use a test database to verify query performance
    }
}
```

### 3. OpenTelemetry Instrumentation for Performance Monitoring

Implement OpenTelemetry instrumentation to properly measure and trace both API request time and database query performance:

```rust
use opentelemetry::{
    trace::{Span, Tracer, TracerProvider},
    global, KeyValue,
};
use opentelemetry_sdk::{
    trace::{self, Sampler},
    Resource,
};
use opentelemetry_otlp::WithExportConfig;

// Setup OpenTelemetry once at application startup
fn setup_telemetry() -> Result<trace::Tracer, Box<dyn std::error::Error>> {
    let tracer = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint("http://localhost:4317"),
        )
        .with_trace_config(
            trace::config()
                .with_sampler(Sampler::AlwaysOn)
                .with_resource(Resource::new(vec![
                    KeyValue::new("service.name", "rsky-pds"),
                    KeyValue::new("service.version", env!("CARGO_PKG_VERSION")),
                ])),
        )
        .install_batch(opentelemetry_sdk::runtime::Tokio)?;
    
    Ok(tracer)
}

// Enhanced list_repos_by_collection with OpenTelemetry tracing
#[tracing::instrument(skip_all, fields(collection = %collection, limit = ?limit, has_cursor = cursor.is_some()))]
#[rocket::get("/xrpc/com.atproto.sync.listReposByCollection?<collection>&<limit>&<cursor>")]
pub async fn list_repos_by_collection(
    collection: Option<String>,
    limit: Option<i64>,
    cursor: Option<String>,
    db: DbConn,
) -> Result<Json<ListReposByCollectionOutput>, ApiError> {
    let tracer = global::tracer("list_repos_by_collection");
    let mut request_span = tracer.start("api.request");
    request_span.set_attribute(KeyValue::new("api.name", "listReposByCollection"));
    request_span.set_attribute(KeyValue::new("collection", collection.clone().unwrap_or_default()));
    
    // Convert and validate the collection
    let collection = match collection {
        Some(c) => {
            let mut validation_span = tracer.start("validate.collection");
            let result = match NsidCollection::new(c) {
                Ok(nsid) => {
                    validation_span.set_attribute(KeyValue::new("validation.success", true));
                    nsid
                },
                Err(e) => {
                    validation_span.set_attribute(KeyValue::new("validation.success", false));
                    validation_span.set_attribute(KeyValue::new("error.type", "InvalidNsid"));
                    validation_span.end();
                    request_span.end();
                    return Err(e);
                },
            };
            validation_span.end();
            result
        },
        None => {
            request_span.set_attribute(KeyValue::new("error.type", "MissingCollection"));
            request_span.end();
            return Err(ApiError::InvalidRequest("Collection query parameter must be set".to_string()));
        },
    };

    // Process cursor
    let cursor_data = if let Some(cursor_str) = cursor {
        let mut cursor_span = tracer.start("parse.cursor");
        let result = match decode_cursor(&cursor_str) {
            Ok(data) => {
                cursor_span.set_attribute(KeyValue::new("cursor.valid", true));
                Some(data)
            },
            Err(_) => {
                cursor_span.set_attribute(KeyValue::new("cursor.valid", false));
                cursor_span.end();
                request_span.end();
                return Err(ApiError::InvalidRequest("Invalid cursor format".to_string()));
            },
        };
        cursor_span.end();
        result
    } else {
        None
    };

    // Execute the query with tracing
    let mut query_span = tracer.start("db.query");
    query_span.set_attribute(KeyValue::new("db.operation", "list_repos_by_collection"));
    query_span.set_attribute(KeyValue::new("db.collection", collection.as_str()));
    query_span.set_attribute(KeyValue::new("db.has_cursor", cursor_data.is_some()));
    
    let result = match inner_list_repos_by_collection(collection, limit, cursor_data, db).await {
        Ok(res) => {
            let repos_count = res.repos.len();
            query_span.set_attribute(KeyValue::new("db.result_count", repos_count as i64));
            query_span.set_attribute(KeyValue::new("db.has_more", res.cursor.is_some()));
            query_span.end();
            request_span.end();
            Ok(res)
        },
        Err(error) => {
            query_span.set_attribute(KeyValue::new("error.type", "DatabaseError"));
            query_span.set_attribute(KeyValue::new("error.message", error.to_string()));
            query_span.end();
            request_span.end();
            Err(error)
        }
    };
    
    result
}

// Enhanced inner query function that records detailed metrics on the database operation
async fn inner_list_repos_by_collection(
    collection: NsidCollection,
    limit: Option<i64>,
    cursor_data: Option<CursorData>,
    db: DbConn,
) -> Result<Json<ListReposByCollectionOutput>, ApiError> {
    let tracer = global::tracer("list_repos_by_collection");
    
    // Default limit to 500 if not provided, max 2000
    let limit = limit.unwrap_or(500).min(2000).max(1);

    // Query records table to find DIDs that have the specified collection
    use crate::schema::pds::record::dsl as RecordSchema;
    
    // Define a result struct for SQL query
    #[derive(QueryableByName, Debug)]
    struct DidResult {
        #[diesel(sql_type = Text)]
        did: String,
    }
    
    // Detailed DB span for query execution
    let mut db_span = tracer.start("db.execute_query");
    db_span.set_attribute(KeyValue::new("db.collection", collection.as_str()));
    db_span.set_attribute(KeyValue::new("db.limit", limit));
    
    if let Some(data) = &cursor_data {
        db_span.set_attribute(KeyValue::new("db.cursor.did", data.did.clone()));
        db_span.set_attribute(KeyValue::new("db.cursor.last_seen_id", data.last_seen_id));
    }
    
    let query_start = std::time::Instant::now();
    
    let repos = match db.run(move |conn| -> Result<Vec<String>, diesel::result::Error> {
        if let Some(cursor_data) = &cursor_data {
            // Use a raw SQL query with cursor pagination
            let query = diesel::sql_query(format!(
                "SELECT DISTINCT did FROM pds.record 
                WHERE collection = $1 
                AND (did > $2 OR (did = $2 AND CTID::text::point[0]::int8 > $3))
                ORDER BY did ASC
                LIMIT $4"
            ))
            .bind::<Text, _>(collection.as_str())
            .bind::<Text, _>(&cursor_data.did)
            .bind::<diesel::sql_types::BigInt, _>(cursor_data.last_seen_id)
            .bind::<diesel::sql_types::BigInt, _>(limit + 1);
            
            // Execute and map results
            let results: Vec<DidResult> = query.load(conn)?;
            Ok(results.into_iter().map(|r| r.did).collect::<Vec<String>>())
        } else {
            // Standard query without cursor
            let results: Vec<String> = RecordSchema::record
                .select(RecordSchema::did)
                .filter(RecordSchema::collection.eq(collection.as_str()))
                .distinct()
                .order_by(RecordSchema::did.asc())
                .limit(limit + 1)
                .load(conn)?;
            
            Ok(results)
        }
    }).await {
        Ok(repos) => {
            let query_duration = query_start.elapsed();
            db_span.set_attribute(KeyValue::new("db.query_time_ms", query_duration.as_millis() as i64));
            db_span.set_attribute(KeyValue::new("db.rows_returned", repos.len() as i64));
            repos
        },
        Err(error) => {
            let query_duration = query_start.elapsed();
            db_span.set_attribute(KeyValue::new("db.query_time_ms", query_duration.as_millis() as i64));
            db_span.set_attribute(KeyValue::new("db.error", true));
            db_span.set_attribute(KeyValue::new("db.error_message", error.to_string()));
            db_span.end();
            return Err(ApiError::InvalidRequest(format!("Failed to retrieve repository data: database error")));
        }
    };
    
    // Process results
    let has_more = repos.len() > limit as usize;
    let repos_to_return = if has_more {
        repos[..limit as usize].to_vec()
    } else {
        repos
    };

    // Generate cursor
    let next_cursor = if has_more && !repos_to_return.is_empty() {
        let mut cursor_span = tracer.start("generate.cursor");
        let last_did = repos_to_return.last().unwrap();
        
        cursor_span.set_attribute(KeyValue::new("cursor.last_did", last_did.clone()));
        
        let cursor_data = CursorData {
            last_seen_id: chrono::Utc::now().timestamp(),
            did: last_did.clone(),
        };
        
        let result = match encode_cursor(&cursor_data) {
            Ok(cursor) => {
                cursor_span.set_attribute(KeyValue::new("cursor.success", true));
                cursor_span.set_attribute(KeyValue::new("cursor.length", cursor.len() as i64));
                Some(cursor)
            },
            Err(_) => {
                cursor_span.set_attribute(KeyValue::new("cursor.success", false));
                None
            }
        };
        
        cursor_span.end();
        result
    } else {
        None
    };
    
    db_span.end();
    
    // Format response
    let result = ListReposByCollectionOutput {
        cursor: next_cursor,
        repos: repos_to_return.iter().map(|did| Repo { did: did.clone() }).collect(),
    };

    Ok(Json(result))
}

// For load testing and benchmarking, we'd set up scripts that collect and analyze the telemetry data
#[cfg(test)]
mod benchmarks {
    use super::*;
    use reqwest::Client;
    
    // Load test that invokes the API and measures performance
    async fn load_test_collection_query() -> Result<(), Box<dyn std::error::Error>> {
        let client = Client::new();
        
        // Test with collection index only
        for _ in 0..100 {
            let response = client
                .get("http://localhost:8000/xrpc/com.atproto.sync.listReposByCollection")
                .query(&[("collection", "app.bsky.feed.post")])
                .send()
                .await?;
                
            println!("Status: {}", response.status());
        }
        
        // After applying compound index, run the same test again to compare results
        
        Ok(())
    }
}

// Structured metrics for comparing different index approaches
#[derive(Debug, Serialize)]
struct IndexPerformanceReport {
    test_date: chrono::DateTime<chrono::Utc>,
    collection_tested: String,
    sample_size: usize,
    
    // Collection index only
    collection_index_metrics: QueryMetrics,
    
    // With compound index
    compound_index_metrics: Option<QueryMetrics>,
    
    // Improvement percentage if both tests were run
    time_improvement_percent: Option<f64>,
}

#[derive(Debug, Serialize)]
struct QueryMetrics {
    avg_query_time_ms: f64,
    median_query_time_ms: f64,
    p95_query_time_ms: f64,
    p99_query_time_ms: f64,
    min_query_time_ms: f64,
    max_query_time_ms: f64,
    
    avg_result_count: f64,
    total_rows_processed: usize,
}
```

The OpenTelemetry implementation would allow:

1. **Complete Request Tracing:** Track the entire request lifecycle from validation to response
3. **Detailed Query Performance:** Measure specific database query execution time
4. **Cursor Performance:** Analyze cursor encoding/decoding overhead
5. **Component-Level Metrics:** Break down performance by validation, query execution, and response formatting
6. **Distributed Tracing:** Connect performance across multiple services when using microservice architecture
7. **Error Rate Analysis:** Track and categorize errors with proper attribution
8. **Visualization:** Connect with tools like Jaeger or Zipkin for visual performance analysis

This approach is much more comprehensive than simple benchmarks and provides actionable data about real-world performance characteristics.


### 4. Cache Layer

Add a caching layer for frequently requested collections:

```rust
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

// Cache implementation
struct CollectionCache {
    cache: Arc<RwLock<HashMap<String, Vec<String>>>>,
    ttl: std::time::Duration,
}

impl CollectionCache {
    // Cache methods
    async fn get(&self, collection: &str) -> Option<Vec<String>> {
        // Get from cache
    }
    
    async fn set(&self, collection: &str, dids: Vec<String>) {
        // Set in cache
    }
    
    async fn invalidate(&self, collection: &str) {
        // Invalidate cache
    }
}
```

These future improvements would enhance the performance, reliability, and maintainability of the `com.atproto.sync.listReposByCollection` implementation.

afbase commented:
Will break this up into several commits.  One with listreposbycollection  example to serve as implementation inspiration.  The others will contain sync1.1 changes reflected for the PDS

afbase commented:
closing for https://github.com/blacksky-algorithms/rsky/pull/88/files
